enhancements/authentication/improved-login-workflow.md

```diff
@@ -0,0 +1,348 @@
+---
+title: oc-authorization-grant-login
+authors:
+  - "@arjunrn"
+reviewers:
+  - "@stlaz"
+  - "@s-urbaniak"
+approvers:
+  - TBD
+api-approvers: # in case of new or modified APIs or API extensions (CRDs, aggregated apiservers, webhooks, finalizers)
+  - NA
+creation-date: 2021-11-08
+last-updated: 2021-11-08
+status: provisional
+---
+
+# oc Authorization Grant Login
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+The current CLI login process involves multiple interactive steps where the user has to enter credentials multiple
+times. This process can be simplified so that the `oc` tool can be used to login directly without having to first login
+to the cluster console, the OAuth server and finally copying the login command. The `oc` tool can be modified to act as
+a public client to the Openshift OAuth server and received an access token
+through [Authorization Grant](https://datatracker.ietf.org/doc/html/rfc6749#section-1.3) flow
+and [PKCE](https://datatracker.ietf.org/doc/html/rfc7636).
+
+## Motivation
+
+Currently, when logging in with the `oc` client a user has to perform the following steps:
+
+1. Open the console in the browser.
+2. Login to console with the configured identity provider.
+3. Click on the **"Copy Login Command"** button which redirects the page to the OAuth application.
+4. Perform login again like in step 2.
+5. Copy the displayed login command.
+6. Run the copied command to configure the client to configure the client and access the cluster.
+
+This process can be simplified to the following steps:
+
+1. User runs `oc login --auto $CLUSTER_API_URL`
+2. `oc` starts a webserver to receive a callback from the OAuth server. It then opens the authorization endpoint with a
+   PKCE code challenge.
+3. The OAuth application performs the authentication either directly or by redirecting the user.
+4. Once authentication is successful the OAuth application redirects the user back to the callback URL of `oc` along
```

Comment:
does the oauth server allow us to redirect back to localhost?

Reply:
Yes, the redirect URI is actually supplied by the client. This is verified by the OAuth server to be one of the previously registered callback URIs of the client.

Reply:
OAuth2 supports Native Apps. They have localhost-like URIs.

The callback URIs can be a private-use URI scheme (`com.example.app`), claimed "https" scheme URI (`https://app.example.com/oauth2redirect/example-provider` or loopback redirect (`http://127.0.0.1:51004/oauth2redirect/example-provider`). [RFC 8252](https://datatracker.ietf.org/doc/html/rfc8252#section-7).

The IdP were we are registered must be also happy to store a redirect URI that contains `localhost:port`.

Reply:
interesting, so all IdPs must be reconfigured for this feature to work?

Reply:
No, the IDPs don't have to be reconfigured. The `OAuthClient` resource has the [redirectURIs](https://docs.openshift.com/container-platform/4.9/rest_api/oauth_apis/oauthclient-oauth-openshift-io-v1.html#oauthclient-oauth-openshift-io-v1) which needs to be set to to the localhost URL.